README
NeuG - a true random number generator implementation (for STM32F103)

							  Version 1.0.6
							     2017-10-11
						           Niibe Yutaka
				      Free Software Initiative of Japan

What's NeuG?
============

NeuG is a set of routines of random number generator (RNG) which is
based on physical noise.  It supports STM32F103.  It can be stand
alone USB RNG device (with main routine), too.

The name comes from Japanized English word "noidgy" (from English word
"noisy"), where many Japanese (including me) don't distinguish
pronunciations of "gee" and "zee".  NeuG includes my important letters
of "g", "n", and "u", and the word "neu" (German spelling of "new").

My primary intention was to incorporate NeuG routines into Gnuk for
random number generation, but the stand alone version could be useful
too.

Gnuk was named after my son who loved the pacifier when he was baby.
NeuG was named after my daughter, but I don't say she is noisy.


Release notes
=============

This is the sixth minor release of NeuG, after 1.0.

Basic features (generating random numbers) are stable, but newly added
things like Fraucheky (USB Mass Storage Class) support and reGNUal
(firmware upgrad) support should be considered experimental.

Note that you need Chopstx (the thread library) as external source
code.  If you get the source code by Git, Chopstx is a submodule of
NeuG.

You can test how the NeuG stand-alone device works on GNU/Linux,
without real hardware, by USBIP emulation (--target=GNU_LINUX).
Please note that this is only for testing.  Use of its generated
randomness is not recommended, because of no entropy source.


FAQ
===

Q0: How NeuG device is good?
A0: I believe it's good enough if we compare to other hardware RNGs.
    I evaluated it with rngtest of RNG-tools, NIST STS test suite and
    Dieharder.  See the directory neug/test-results/.
A0-dash: For better entropy device with embedded test, you could get
         EntropyKey.  See http://www.entropykey.co.uk/
A0-double-dash: STM32F2xx and STM32F4xx have built-in TRNG, it would
                be better for you (although the quality of randomness
                looks not that good).

Q1: How fast is NeuG device?
A1: It's about 80 kB/sec for conditioned output (by SHA-256), and 
    about 270 kB/sec for CRC-32 filtered output (kB = 1000 byte).

Q2: Should we check condition of noise sources?
A2: Yes, we should.  Three continuous tests are implemented, following
    (Draft of) NIST SP 800-90B.  Those are Repetition Count Test,
    Adaptive Proportion Test (for 64 samples), and another Adaptive
    Proportion Test (for 4096 samples).  When it detects an error (it
    is really rare, but it could occur even for normal condition), the
    generation of random bits restart again.

Q3: Conditioning with SHA-256 sounds over-kill.  Why not simpler?
A3: It is because NIST SP 800-90B mandates something like that for
    "full entropy source".  If your usage is as an entropy source for
    RNG-tools to feed entropy to your kernel, or use for computer
    simulations, I think that CRC-32 filtered output would be good
    enough.

Q4: Configuring by 'stty' doesn't work, what's up?
A4: Common problem on GNU/Linux is the "modemmanager" try to control
    /dev/ttyACM0, which interferes.  Please consider removing
    modemmanager from your system or configure it to ignore NeuG
    device.  For detail, please see the FST-01 support pages on the
    web.


Targets
=======

FST-01, Nitrokey-Start, Olimex STM32-H103, STM32 part of STM8S
Discovery Kit, STBee Mini, STBee, STM32 Primer 2, CQ STARM, ST Dongle
and STM32 Nucleo F103 are supported.


Build system and Host system
============================

Makefile is written for GNU make.  You need Bash 4.x for configure.

If your bash is not installed as /bin/bash, you need to run configure
script prepending 'bash' before './configure'.

Some tools are written in Python.  If your Python is not installed as
/usr/bin/python, please prepend 'python' for your command invocation.
Python 2.7 and PyUSB 0.4.3 is assumed.  I also use Python 3.5 and
PyUSB 1.0.0.


Souce code
==========

NeuG source code is under src/ directory.

Note that SHA-256 hash function implementation, src/sha256.c, is based
on the original implementation by Dr. Brian Gladman.  See:

  http://brg.a2hosted.com//oldsite/cryptography_technology/sha/index.php
(was at:
 http://gladman.plushost.co.uk/oldsite/cryptography_technology/sha/index.php)


License
=======

It is distributed under GNU General Public Licence version 3 or later
(GPLv3+).  Please see src/COPYING.


External source code
====================

NeuG is distributed with external source code.

* chopstx/ -- Chopstx, the RT Thread Library

* [optional] fraucheky/ -- Fraucheky, the GPL container

Those are available at:

  https://anonscm.debian.org/cgit/gnuk/chopstx/


USB vendor ID and product ID (USB device ID)
============================================

When you have a vender ID and assign a product ID for NeuG, edit the
file NEUG_USB_DEVICE_ID and add an entry for yours.  In this case,
please contact Niibe, so that it is listed to the file in the official
release of the source code.

When you are modifing NeuG and installing the binary to device, you
should replace the vendor string to yours, so that users can see it's
not by original vendor, and it is modified version.

FSIJ allows you to use USB device ID of FSIJ (234b:0001) for devices
with standalone NeuG under one of following conditions:

  * For everyone for experimental purpose:

    - You must not distribute a binary with FSIJ's USB device ID, but
      must use the binary by yourself only for your experiment.  Note
      that "Distributing binary" includes distributing a device which
      holds the binary.

  * For individuals:

    - No additional conditions.

FSIJ could give companies or business entities "second source
manufacturer" license to use USB device ID of FSIJ for devices with
unmodified version of NeuG, provided they support Free Software and
respect users' freedom for computing.  Please ask FSIJ for the
license.

Otherwise, companies which want to distribute NeuG devices, please use
your own USB vendor ID and product ID.  Please replace vendor string
and possibly product string to yours, when you modify NeuG.


How to compile
==============

You need GNU toolchain and newlib for 'arm-none-eabi' target.

On Debian we can install the packages of gcc-arm-none-eabi,
gdb-arm-none-eabi and its friends.  I'm using:

	binutils-arm-none-eabi	2.28-5+9+b3
	gcc-arm-none-eabi 	15:5.4.1+svn241155-1
	gdb-arm-none-eabi 	7.12-6+9+b2
	libnewlib-arm-none-eabi	2.4.0.20160527-2

Or else, see https://launchpad.net/gcc-arm-embedded for preparation of
GNU Toolchain for 'arm-none-eabi' target.

Change directory to `src':

  $ cd neug-VERSION/src

Then, run `configure':

  $ ./configure --vidpid=<VID:PID>

Here, you need to specify USB vendor ID and product ID.  For FSIJ's,
it's: --vidpid=234b:0001 .  Please read section 'USB vendor ID and
product ID' above.


Type:

  $ make

Then, we will have "neug.elf" under the "build" directory.


How to install
==============

Olimex STM32-H103 board
-----------------------

If you are using Olimex JTAG-Tiny, type following to invoke OpenOCD:

  $ openocd -f interface/ftdi/olimex-jtag-tiny.cfg -f board/olimex_stm32_h103.cfg

Then, with another terminal, type following to write "neug.elf" to Flash ROM:

  $ telnet localhost 4444
  > reset halt
  > flash write_image erase neug.elf
  > reset
  > exit
  $ 


Flying Stone Tiny 01 and STM8S Discovery Kit
--------------------------------------------

If you are using Flying Stone Tiny 01, you need a SWD writer.

OpenOCD 0.6.1 or newer supports ST-Link/V2.  With that we can do:

  $ openocd -f interface/stlink-v2.cfg -f target/stm32f1x_stlink.cfg

Then, with another terminal, type following to write "neug.elf" to Flash ROM:

  $ telnet localhost 4444
  > reset halt
  > flash write_image erase neug.elf
  > reset
  > exit
  $ 

Note that OpenOCD (as of 0.7.0) doesn't support option bytes handling yet.

You can also use tool/stlinkv2.py in Gnuk.


Use of NeuG device
==================

It is USB CDC ACM device.  On GNU/Linux, it can be /dev/ttyACM0 or like.
Before using /dev/ttyACM0, you need to configure its TTY discipline.

  $ stty -F /dev/ttyACM0 -echo raw

Then, you can use output of /dev/ttyACM0.

When you want to get CRC-32 filtered output, you can configure:

  $ stty -F /dev/ttyACM0 parenb parodd

for raw data after filter.  For direct raw data of ADC samples,
configure:

  $ stty -F /dev/ttyACM0 parenb -parodd

And you can get SHA-256 conditioned output by configuring:

  $ stty -F /dev/ttyACM0 -parenb


If your device is configured with Fraucheky, you can invoke Fraucheky
from NeuG by setting the unusual sequence of configurations:

  $ stty -F /dev/ttyACM0 ispeed 110 ospeed 110
  $ stty -F /dev/ttyACM0 ispeed 115200 ospeed 115200


Structure of NeuG
=================

Here is a figure of the circuit.

                                         Noise sources

                              /|<---+--- [ Analog input Vref ]
                16           | |<-+-|--- [ Analog input Temperature Sensor ]
             +---/-[ADC1] <==| |  | |
             |               | |<-+ |
      +-+    |                \|<---+
      | |<---+                |
 +----| |          MUX CTL >--+
 |    | |<---+
 |    +-+    |                /|
 |           |  16           | |<------- [ Analog input 0 ] (pull up to Vdd)
 |           +---/-[ADC2] <==| |
 |                           | |<------- [ Analog input 1 ] (pull up to Vdd)
 |                            \|
 |                            |
 |                 MUX CTL >--+
 |
 +------------------+         <============ (*1)
                    |
                    / 32
                    |
                    | Put 4 times to output 32-bit
                    V
            [ CRC-32 filter ]
                    |
                    |           Put 35 times to output 1120-bit
                    +---------------------------------+         <====== (*2)
                                                      |
                                                      / 32
                                                      |
                                                      V
                                              [ Entropy Buffer ]
                                                      |          
            +--------------+                          |
            |              |                          |
            | Conditioning |            1120          |
            | Component    |<------------/------------+
            |              |
      +-----|  Hash_df     |
      |     |    by        |
      |     |  SHA-256     |
      |     |              |  128
      |     |              |<--/--+
      |     +--------------+      |
      |                           |
      +---------------------------+
      |
      / 256
      |
      v
 Random Number Output <========== (*3)



Specifying by "stty", you can get (*3) with -parenb, (*2) with parenb
parodd, and (*1) with parenb -parodd.

STM32F103 has two built-in A/D converters.  NeuG uses A/D converters'
outputs as entropy sources.  It is considered noise of quantization
error, plus noise from power supply, etc.

We chose four analog input sources of: built-in voltage reference,
temperature sensor and two analog inputs which are pull-up to Vdd.

By a single sampling of two channels, we get 32-bit (not all 32-bit is
valid, as a A/D converter resolution is 12-bit only).  We take four
sampling of combinations: (Vref, IN0), (Temp, IN1), (Vref, IN1), and

NEWS
NeuG NEWS - User visible changes


* Major changes in NeuG 1.0.6

  Released 2017-10-11, by NIIBE Yutaka

** Port to GNU/Linux emulation
We can "run" NeuG stand-alone device on GNU/Linux by emulation through
USBIP.

** Upgrade of Fraucheky
We use Fraucheky 0.4, which can work well with GNU/Linux emulation.

** Upgrade of Chopstx
We use Chopstx 1.5, which supports Blue Pill and the emulation on
GNU/Linux.


* Major changes in NeuG 1.0.5

  Released 2016-10-17, by NIIBE Yutaka

** Upgrade of Chopstx
We use Chopstx 1.2, which supports FST-01G.


* Major changes in NeuG 1.0.4

  Released 2015-09-18, by NIIBE Yutaka

** Upgrade of Chopstx
We use Chopstx 0.10, which supports Nitrokey-Start.

** USB reset handling
Now, USB reset lets NeuG device restart correctly.


* Major changes in NeuG 1.0.3

  Released 2015-08-03, by NIIBE Yutaka

** Fraucheky can be manually invoked
Even if Fraucheky is disabled at start by moving files to DROPHERE,
a user can invoke Fraucheky to see documents, now.

** Upgrade of Chopstx
We use Chopstx 0.08, which supports ST-Dongle and STM32 Nucleo F103.

** Upgrade of Fraucheky
We use Fraucheky 0.3, which can work well with FreeBSD.


* Major changes in NeuG 1.0.2

  Released 2015-07-18, by NIIBE Yutaka

** Upgrade of firmware enhancement
The process is protected by password.

** Upgrade of Chopstx
We use Chopstx 0.07, which supports STM32 Primer2, CQ STARM.


* Major changes in NeuG 1.0.1

  Released 2015-01-29, by NIIBE Yutaka

** Upgrade of Chopstx
We use Chopstx 0.04.

** Upgrade of Fraucheky
We use Fraucheky 0.1.

** Compile options
We use -O3 (with no -Os) for neug.o and sha256.o.


* Major changes in NeuG 1.0

  Released 2013-11-11, by NIIBE Yutaka

** Upgrade of Chopstx
Now, we use Chopstx 0.03.

** Stabilize the upgrade process
For firmware upgrade, it has been unstable somehow and it has been
recommended not to access its stream (/dev/ttyACM0) before running
neug_upgrade.py.  This bug was fixed in 1.0, and it's more stable.

** Add support of Fraucheky
Fraucheky is a GPL container which makes sure to deliver GPL to users.


* Major changes in NeuG 0.11

  Released 2013-08-22, by NIIBE Yutaka

** Replacement of kernel (thread library)
Instead of ChibiOS/RT, we now use Chopstx.

** Improved performance
The output of random numbers got faster than the previous
implementation by 30% or so.

** Unsupported targets
CQ_STARM, STBEE, STBEE Mini, and STM32_PRIMER2 are not
supported in this release, but porting should be easy.


* Major changes in NeuG 0.06

  Released 2013-03-15, by NIIBE Yutaka

** Support for relocatable reGNUal
For firmware upgrade, we can use newer reGNUal (>= 1.0.4).


* Major changes in NeuG 0.05

  Released 2013-02-07, by NIIBE Yutaka

** Improvement on SHA256 calculation
The output of random numbers got faster than the previous
implementation by 2% or so.


* Major changes in NeuG 0.04

  Released 2012-10-30, by NIIBE Yutaka

** Update of ChibiOS/RT
ChibiOS/RT is updated to newer 2.4.3.

** Improved performance
Improvement on generation speed, even when it's _not_ connected to USB
2.0 Hub.

** Health tests run for NEUG_MODE_RAW
Health tests run when NEUG_MODE_RAW (CRC32 filter output).

** More runtime info: tool/neug_check.py 
Now, we can observe max counts of repeat count test, adaptive
propotion test 64, and adaptive propotion test 4096.


* Major changes in NeuG 0.03

  Released 2012-10-09, by NIIBE Yutaka

** Change RNG constuct.
We replace all basic components again, to conform NIST SP 800-90B.
Now, we use two analog inputs use in addition to Vref and Temperature
Sensor.  We take advantage of another A/D convertor, ADC2.

In former versions, we use LSBs of inputs.  From this version, we use
all bits, and put them to CRC32 calculation unit.  That's because we
can get more entropy when using full bits.

The central component, the conditioning component, is the one with
SHA-256, to conform NIST SP 80-90B.

** New ADC (A/D convertor) driver
Now, we use our own ADC driver, so that we can use ADC2 and we can
control ADC for our particular usage (such that put it to CRC32
calculation unit).

** Selection of output
Using "stty" command, we can now select one of three outputs: final
output (conditioned), CRC32 filtered raw data before conditioning, and
raw input data of ADC.

** USB strings for revision detail, configure options, and system service.
USB strings now have more information.  There are revision detail
string, configure options string, system service version string, as
well as vendor string and product string.  These strings could be
examined to check NeuG device.

** New mandatory option '--vidpid' for configure
You must specify USB vendor ID and product ID for Gnuk.
The file NEUG_USB_DEVICE_ID lists valid USB device IDs.

** Health tests
Health tests are implemented to conform NIST SP 800-90B.  There are
three tests, repetition count test, adaptive proportion test 64,
and adaptive proportion test 4096.

** System service blocks at the beginning of flash ROM.
Once flash ROM is protected, first 4-KiB cannot be modified.  Gnuk use
this area for "system service".  NeuG has compatible system service
block as Gnuk.  Note that this area will not be able to be modified by
firmware upgrade (or by any method).

** New tool: neug_upgrade.py
The tool neug_upgrade.py is to send reGNUal to NeuG device.  Then, we
put new NeuG binary (or any binary, say, Gnuk) into the device with
reGNUal.

** New tool: neug_check.py
The tool neug_check.py is to check the device condition.  You can see
error counters of NeuG device with this tool.


* Major changes in NeuG 0.02

  Released 2012-08-30, by NIIBE Yutaka

** Change RNG constuct.
We replace all basic components.  Now, we use WELL 512a structure
for Physical-based RNG.

** New USB stack.
Now, we use our own USB stack (taken from Gnuk).



src/random.c
/*
 * random.c - random number generation
 *
 * Copyright (C) 2011, 2012, 2013 Free Software Initiative of Japan
 * Author: NIIBE Yutaka <gniibe@fsij.org>
 *
 * This file is a part of NeuG, a True Random Number Generator
 * implementation based on quantization error of ADC (for STM32F103).
 *
 * NeuG is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * NeuG is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdint.h>
#include <string.h>
#include <chopstx.h>

#include "sys.h"
#include "neug.h"
#include "stm32f103.h"
#include "adc.h"
#include "sha256.h"

chopstx_mutex_t adc_mtx;
chopstx_cond_t adc_cond;
int adc_waiting;
int adc_data_available;

static uint32_t adc_buf[SHA256_BLOCK_SIZE/sizeof (uint32_t)];

static sha256_context sha256_ctx_data;
static uint32_t sha256_output[SHA256_DIGEST_SIZE/sizeof (uint32_t)];

/*
 * To be a full entropy source, the requirement is to have N samples
 * for output of 256-bit, where:
 *
 *      N = (256 * 2) / <min-entropy of a sample>
 *
 * For example, N should be more than 103 for min-entropy = 5.0.
 *
 * On the other hand, in the section 6.2 "Full Entropy Source
 * Requirements", it says:
 *
 *     At least twice the block size of the underlying cryptographic
 *     primitive shall be provided as input to the conditioning
 *     function to produce full entropy output.
 *
 * For us, cryptographic primitive is SHA-256 and its blocksize is
 * 512-bit (64-byte), thus, N >= 128.
 *
 * We chose N=140.  Note that we have "additional bits" of 16-byte for
 * last block (feedback from previous output of SHA-256) to feed
 * hash_df function of SHA-256, together with sample data of 140-byte.
 *
 * N=140 corresponds to min-entropy >= 3.68.
 *
 */
#define NUM_NOISE_INPUTS 140

#define EP_ROUND_0 0 /* initial-five-byte and 3-byte, then 56-byte-input */
#define EP_ROUND_1 1 /* 64-byte-input */
#define EP_ROUND_2 2 /* 17-byte-input */
#define EP_ROUND_RAW      3 /* 32-byte-input */
#define EP_ROUND_RAW_DATA 4 /* 32-byte-input */

#define EP_ROUND_0_INPUTS 56
#define EP_ROUND_1_INPUTS 64
#define EP_ROUND_2_INPUTS 17
#define EP_ROUND_RAW_INPUTS 32
#define EP_ROUND_RAW_DATA_INPUTS 32

static uint8_t ep_round;

/*
 * Hash_df initial string:
 *
 *  1,          : counter = 1
 *  0, 0, 1, 0  : no_of_bits_returned (in big endian)
 */
static void ep_fill_initial_string (void)
{
  adc_buf[0] = 0x01000001; /* Regardless of endian */
  adc_buf[1] = (CRC->DR & 0xffffff00);
}

static void ep_init (int mode)
{
  adc_data_available = 0;

  if (mode == NEUG_MODE_RAW)
    {
      ep_round = EP_ROUND_RAW;
      adc_start_conversion (ADC_CRC32_MODE, adc_buf, EP_ROUND_RAW_INPUTS);
    }
  else if (mode == NEUG_MODE_RAW_DATA)
    {
      ep_round = EP_ROUND_RAW_DATA;
      adc_start_conversion (ADC_SAMPLE_MODE, adc_buf, EP_ROUND_RAW_DATA_INPUTS);
    }
  else
    {
      ep_round = EP_ROUND_0;
      ep_fill_initial_string ();
      adc_start_conversion (ADC_CRC32_MODE,
			    &adc_buf[2], EP_ROUND_0_INPUTS);
    }
}

static void noise_source_continuous_test (uint8_t noise);

static void ep_fill_wbuf (int i, int flip, int test)
{
  uint32_t v = adc_buf[i];

  if (test)
    {
      uint8_t b0, b1, b2, b3;

      b3 = v >> 24;
      b2 = v >> 16;
      b1 = v >> 8;
      b0 = v;

      noise_source_continuous_test (b0);
      noise_source_continuous_test (b1);
      noise_source_continuous_test (b2);
      noise_source_continuous_test (b3);
    }

  if (flip)
    v = __builtin_bswap32 (v);

  sha256_ctx_data.wbuf[i] = v;
}

/* Here assumes little endian architecture.  */
static int ep_process (int mode)
{
  int i, n;

  if (ep_round == EP_ROUND_RAW)
    {
      for (i = 0; i < EP_ROUND_RAW_INPUTS / 4; i++)
	ep_fill_wbuf (i, 0, 1);

      ep_init (mode);
      return EP_ROUND_RAW_INPUTS / 4;
    }
  else if (ep_round == EP_ROUND_RAW_DATA)
    {
      for (i = 0; i < EP_ROUND_RAW_DATA_INPUTS / 4; i++)
	ep_fill_wbuf (i, 0, 0);

      ep_init (mode);
      return EP_ROUND_RAW_DATA_INPUTS / 4;
    }

  if (ep_round == EP_ROUND_0)
    {
      for (i = 0; i < 64 / 4; i++)
	ep_fill_wbuf (i, 1, 1);

      adc_start_conversion (ADC_CRC32_MODE, adc_buf, EP_ROUND_1_INPUTS);
      sha256_start (&sha256_ctx_data);
      sha256_process (&sha256_ctx_data);
      ep_round++;
      return 0;
    }
  else if (ep_round == EP_ROUND_1)
    {
      for (i = 0; i < 64 / 4; i++)
	ep_fill_wbuf (i, 1, 1);

      adc_start_conversion (ADC_CRC32_MODE, adc_buf, EP_ROUND_2_INPUTS);
      sha256_process (&sha256_ctx_data);
      ep_round++;
      return 0;
    }
  else
    {
      for (i = 0; i < (EP_ROUND_2_INPUTS + 3) / 4; i++)
	ep_fill_wbuf (i, 0, 1);

      n = SHA256_DIGEST_SIZE / 2;
      ep_init (NEUG_MODE_CONDITIONED); /* The three-byte is used here.  */
      memcpy (((uint8_t *)sha256_ctx_data.wbuf)
	      + ((NUM_NOISE_INPUTS+5)%SHA256_BLOCK_SIZE),
	      sha256_output, n); /* Don't use the last three-byte.  */
      sha256_ctx_data.total[0] = 5 + NUM_NOISE_INPUTS + n;
      sha256_finish (&sha256_ctx_data, (uint8_t *)sha256_output);
      return SHA256_DIGEST_SIZE / sizeof (uint32_t);
    }
}


static const uint32_t *ep_output (int mode)
{
  if (mode)
    return sha256_ctx_data.wbuf;
  else
    return sha256_output;
}

#define REPETITION_COUNT           1
#define ADAPTIVE_PROPORTION_64     2
#define ADAPTIVE_PROPORTION_4096   4

uint8_t neug_err_state;
uint16_t neug_err_cnt;
uint16_t neug_err_cnt_rc;
uint16_t neug_err_cnt_p64;
uint16_t neug_err_cnt_p4k;

uint16_t neug_rc_max;
uint16_t neug_p64_max;
uint16_t neug_p4k_max;

#include "board.h"

static void noise_source_cnt_max_reset (void)
{
  neug_err_cnt = neug_err_cnt_rc = neug_err_cnt_p64 = neug_err_cnt_p4k = 0;
  neug_rc_max = neug_p64_max = neug_p4k_max = 0;
}

static void noise_source_error_reset (void)
{
  neug_err_state = 0;
}

static void noise_source_error (uint32_t err)
{
  neug_err_state |= err;
  neug_err_cnt++;

  if ((err & REPETITION_COUNT))
    neug_err_cnt_rc++;
  if ((err & ADAPTIVE_PROPORTION_64))
    neug_err_cnt_p64++;
  if ((err & ADAPTIVE_PROPORTION_4096))
    neug_err_cnt_p4k++;
}

/*
 * For health tests, we assume that the device noise source has
 * min-entropy >= 4.2.  Observing raw data stream (before CRC-32) has
 * more than 4.2 bit/byte entropy.  When the data stream after CRC-32
 * filter will be less than 4.2 bit/byte entropy, that must be
 * something wrong.  Note that even we observe < 4.2, we still have
 * some margin, since we use NUM_NOISE_INPUTS=140.
 *
 */

/* Cuttoff = 9, when min-entropy = 4.2, W= 2^-30 */
/* ceiling of (1+30/4.2) */
#define REPITITION_COUNT_TEST_CUTOFF 9

static uint8_t rct_a;
static uint8_t rct_b;

static void repetition_count_test (uint8_t sample)
{
  if (rct_a == sample)
    {
      rct_b++;
      if (rct_b >= REPITITION_COUNT_TEST_CUTOFF)
	noise_source_error (REPETITION_COUNT);
      if (rct_b > neug_rc_max)
	neug_rc_max = rct_b;
   }
  else
    {
      rct_a = sample;
      rct_b = 1;
    }
}

/* Cuttoff = 18, when min-entropy = 4.2, W= 2^-30 */
/* With R, qbinom(1-2^-30,64,2^-4.2) */
#define ADAPTIVE_PROPORTION_64_TEST_CUTOFF 18

static uint8_t ap64t_a;
static uint8_t ap64t_b;
static uint8_t ap64t_s;

static void adaptive_proportion_64_test (uint8_t sample)
{
  if (ap64t_s >= 64)
    {
      ap64t_a = sample;
      ap64t_s = 0;
      ap64t_b = 0;
    }
  else
    {
      ap64t_s++;
      if (ap64t_a == sample)
	{
	  ap64t_b++;
	  if (ap64t_b > ADAPTIVE_PROPORTION_64_TEST_CUTOFF)
	    noise_source_error (ADAPTIVE_PROPORTION_64);
	  if (ap64t_b > neug_p64_max)
	    neug_p64_max = ap64t_b;
	}
    }
}

/* Cuttoff = 315, when min-entropy = 4.2, W= 2^-30 */
/* With R, qbinom(1-2^-30,4096,2^-4.2) */
#define ADAPTIVE_PROPORTION_4096_TEST_CUTOFF 315

static uint8_t ap4096t_a;
static uint16_t ap4096t_b;
static uint16_t ap4096t_s;

static void adaptive_proportion_4096_test (uint8_t sample)
{
  if (ap4096t_s >= 4096)
    {
      ap4096t_a = sample;
      ap4096t_s = 0;
      ap4096t_b = 0;
    }
  else
    {
      ap4096t_s++;
      if (ap4096t_a == sample)
	{
	  ap4096t_b++;
	  if (ap4096t_b > ADAPTIVE_PROPORTION_4096_TEST_CUTOFF)
	    noise_source_error (ADAPTIVE_PROPORTION_4096);
	  if (ap4096t_b > neug_p4k_max)
	    neug_p4k_max = ap4096t_b;
	}
    }
}

static void noise_source_continuous_test (uint8_t noise)
{
  repetition_count_test (noise);
  adaptive_proportion_64_test (noise);
  adaptive_proportion_4096_test (noise);
}

/*
 * Ring buffer, filled by generator, consumed by neug_get routine.
 */
struct rng_rb {
  uint32_t *buf;
  chopstx_mutex_t m;
  chopstx_cond_t data_available;
  chopstx_cond_t space_available;
  uint8_t head, tail;
  uint8_t size;
  unsigned int full :1;
  unsigned int empty :1;
};

static void rb_init (struct rng_rb *rb, uint32_t *p, uint8_t size)
{
  rb->buf = p;
  rb->size = size;
  chopstx_mutex_init (&rb->m);
  chopstx_cond_init (&rb->data_available);
  chopstx_cond_init (&rb->space_available);
  rb->head = rb->tail = 0;
  rb->full = 0;
  rb->empty = 1;
}

static void rb_add (struct rng_rb *rb, uint32_t v)
{
  rb->buf[rb->tail++] = v;
  if (rb->tail == rb->size)
    rb->tail = 0;
  if (rb->tail == rb->head)
    rb->full = 1;
  rb->empty = 0;
}

static uint32_t rb_del (struct rng_rb *rb)
{
  uint32_t v = rb->buf[rb->head++];

  if (rb->head == rb->size)
    rb->head = 0;
  if (rb->head == rb->tail)
    rb->empty = 1;
  rb->full = 0;

  return v;
}

uint8_t neug_mode;
static int rng_should_terminate;
static chopstx_t rng_thread;


/**
 * @brief Random number generation thread.
 */
static void *
rng (void *arg)
{
  struct rng_rb *rb = (struct rng_rb *)arg;

  rng_should_terminate = 0;
  chopstx_mutex_init (&adc_mtx);
  chopstx_cond_init (&adc_cond);

  /* Enable ADCs */
  adc_start ();

  ep_init (NEUG_MODE_CONDITIONED);

  while (!rng_should_terminate)
    {
      int n;
      int mode = neug_mode;

      chopstx_mutex_lock (&adc_mtx);
      if (!adc_data_available)
	{
	  adc_waiting = 1;
	  chopstx_cond_wait (&adc_cond, &adc_mtx);
	  adc_waiting = 0;
	}
      adc_data_available = 0;
      chopstx_mutex_unlock (&adc_mtx);

      if ((n = ep_process (mode)))
	{
	  int i;
	  const uint32_t *vp;

	  if (neug_err_state != 0
	      && (mode == NEUG_MODE_CONDITIONED || mode == NEUG_MODE_RAW))
	    {
	      /* Don't use the result and do it again.  */
	      noise_source_error_reset ();
	      continue;
	    }

	  vp = ep_output (mode);

	  chopstx_mutex_lock (&rb->m);
	  while (rb->full)
	    chopstx_cond_wait (&rb->space_available, &rb->m);

	  for (i = 0; i < n; i++)
	    {
	      rb_add (rb, *vp++);
	      if (rb->full)
		break;
	    }

	  chopstx_cond_signal (&rb->data_available);
	  chopstx_mutex_unlock (&rb->m);
	}
    }

  adc_stop ();

  return NULL;
}

static struct rng_rb the_ring_buffer;

extern uint8_t __process2_stack_base__, __process2_stack_size__;
const uint32_t __stackaddr_rng = (uint32_t)&__process2_stack_base__;
const size_t __stacksize_rng = (size_t)&__process2_stack_size__;
#define PRIO_RNG 2

/**
 * @brief Initialize NeuG.
 */
void
neug_init (uint32_t *buf, uint8_t size)
{
  const uint32_t *u = (const uint32_t *)unique_device_id ();
  struct rng_rb *rb = &the_ring_buffer;
  int i;

  RCC->AHBENR |= RCC_AHBENR_CRCEN;
  CRC->CR = CRC_CR_RESET;

  /*
   * This initialization ensures that it generates different sequence
   * even if all physical conditions are same.
   */
  for (i = 0; i < 3; i++)
    CRC->DR = *u++;

  neug_mode = NEUG_MODE_CONDITIONED;
  rb_init (rb, buf, size);

  rng_thread = chopstx_create (PRIO_RNG, __stackaddr_rng, __stacksize_rng,
			       rng, rb);
}

/**
 * @breif Flush random bytes.
 */
void
neug_flush (void)
{
  struct rng_rb *rb = &the_ring_buffer;

  chopstx_mutex_lock (&rb->m);
  while (!rb->empty)
    (void)rb_del (rb);
  chopstx_cond_signal (&rb->space_available);
  chopstx_mutex_unlock (&rb->m);
}


/**
 * @brief  Wakes up RNG thread to generate random numbers.
 */
void
neug_kick_filling (void)
{
  struct rng_rb *rb = &the_ring_buffer;

  chopstx_mutex_lock (&rb->m);
  if (!rb->full)
    chopstx_cond_signal (&rb->space_available);
  chopstx_mutex_unlock (&rb->m);
}

/**
 * @brief  Get random word (32-bit) from NeuG.
 * @detail With NEUG_KICK_FILLING, it wakes up RNG thread.
 *         With NEUG_NO_KICK, it doesn't wake up RNG thread automatically,
 *         it is needed to call neug_kick_filling later.
 */
uint32_t
neug_get (int kick)
{
  struct rng_rb *rb = &the_ring_buffer;
  uint32_t v;

  chopstx_mutex_lock (&rb->m);
  while (rb->empty)
    chopstx_cond_wait (&rb->data_available, &rb->m);
  v = rb_del (rb);
  if (kick)
    chopstx_cond_signal (&rb->space_available);
  chopstx_mutex_unlock (&rb->m);

  return v;
}

int
neug_get_nonblock (uint32_t *p)
{
  struct rng_rb *rb = &the_ring_buffer;
  int r = 0;

  chopstx_mutex_lock (&rb->m);
  if (rb->empty)
    {
      r = -1;
      chopstx_cond_signal (&rb->space_available);
    }
  else
    *p = rb_del (rb);
  chopstx_mutex_unlock (&rb->m);

  return r;
}

int neug_consume_random (void (*proc) (uint32_t, int))
{
  int i = 0;
  struct rng_rb *rb = &the_ring_buffer;

  chopstx_mutex_lock (&rb->m);
  while (!rb->empty)
    {
      uint32_t v;

      v = rb_del (rb);
      proc (v, i);
      i++;
    }
  chopstx_cond_signal (&rb->space_available);
  chopstx_mutex_unlock (&rb->m);

  return i;
}

void
neug_wait_full (void)
{
  struct rng_rb *rb = &the_ring_buffer;

  chopstx_mutex_lock (&rb->m);
  while (!rb->full)
    chopstx_cond_wait (&rb->data_available, &rb->m);
  chopstx_mutex_unlock (&rb->m);
}

void
neug_fini (void)
{
  rng_should_terminate = 1;
  neug_get (1);
  chopstx_join (rng_thread, NULL);
}

void
neug_mode_select (uint8_t mode)
{
  if (neug_mode == mode)
    return;

  neug_wait_full ();

  chopstx_mutex_lock (&adc_mtx);
  while (adc_waiting == 0)
    {
      chopstx_mutex_unlock (&adc_mtx);
      chopstx_usec_wait (1000);
      chopstx_mutex_lock (&adc_mtx);
    }
  chopstx_mutex_unlock (&adc_mtx);

  ep_init (mode);
  noise_source_cnt_max_reset ();
  neug_mode = mode;
  neug_flush ();
}
